### ISSUE TITLE
Suspected Issue in categorical_crossentropy

### ISSUE THREAD

**Issue by rakeshbarman2790** [Theano/Theano#6819]
Discrepancy in categorical_crossentropy Loss Between Theano and TensorFlow
When computing the categorical_crossentropy loss, a significant difference in results is observed between Theano (1.0.4) and TensorFlow (2.0.0). This discrepancy points to a potential inconsistency in how the two libraries handle loss calculations.

Code Comparison
Theano Implementation
python
Copy code
import numpy as np
from theano import tensor as T

# Define true and predicted values
y_true = np.array([0., 1., 0.], dtype=np.float32)
y_pred = np.array([0.9999999, 0.9999999, 0.0000001], dtype=np.float32)

# Calculate categorical crossentropy
res = T.nnet.categorical_crossentropy(y_pred, y_true)

# Print the result
print("Theano loss =", res.eval())
# Output: Theano loss = 1.192093e-07
TensorFlow Implementation
python
Copy code
import numpy as np
import tensorflow.keras.losses as keras_losses

# Define true and predicted values
y_true = np.array([0., 1., 0.], dtype=np.float32)
y_pred = np.array([0.9999999, 0.9999999, 0.0000001], dtype=np.float32)

# Calculate categorical crossentropy
res = keras_losses.categorical_crossentropy(y_true, y_pred)

# Print the result
print("TensorFlow loss =", res.numpy())
# Output: TensorFlow loss = 0.69314724
Observations
Theano Result:
The computed loss is 1.192093e-07, which is almost zero, indicating that Theano interprets the predictions as near-perfect.

TensorFlow Result:
The computed loss is 0.69314724, which is a substantial value, suggesting TensorFlow penalizes the predictions for not being sharply aligned with the true labels.

Root Cause of the Discrepancy
Logarithmic Stabilization:
TensorFlow applies a stabilization mechanism, clipping the predicted values (y_pred) to a small range (e.g., [1e-7, 1 - 1e-7]) to prevent numerical issues like log(0). Theano does not apply such clipping by default.

Loss Formula:
The categorical_crossentropy loss is defined as:

Loss
=
−
∑
𝑖
𝑦
true
,
𝑖
⋅
log
⁡
(
𝑦
pred
,
𝑖
)
Loss=− 
i
∑
​
 y 
true,i
​
 ⋅log(y 
pred,i
​
 )
In TensorFlow, y_pred is clipped, which ensures a consistent penalty even for predictions that are extremely close to 1.
In Theano, when y_pred values are near 1 or 0, the absence of clipping results in near-zero penalties, leading to inconsistencies.
Precision Differences:
Theano's result may also reflect numerical precision differences in handling extremely small values.

Suggested Fix for Theano
To make Theano's behavior consistent with TensorFlow, manually clip the predicted values:

python
Copy code
from theano import tensor as T

# Clip predicted values to prevent numerical issues
epsilon = 1e-7
y_pred_clipped = T.clip(y_pred, epsilon, 1 - epsilon)

# Recalculate categorical crossentropy with clipped values
res = T.nnet.categorical_crossentropy(y_pred_clipped, y_true)
print("Corrected Theano loss =", res.eval())
Conclusion
The discrepancy arises from the absence of input clipping in Theano's categorical_crossentropy implementation. TensorFlow's use of clipping ensures numerical stability and consistent penalties for imperfect predictions. To avoid such inconsistencies, users of Theano should manually clip predicted values before calculating the loss.
